src/internal/_eqDeep.js

```diff
@@ -0,0 +1,64 @@
+var eq = require('../eq');
+var keys = require('../keys');
+var type = require('../type');
+
+// The algorithm used to handle cyclic structures is
+// inspired by underscore's isEqual
+module.exports = function _eqDeep(a, b, stackA, stackB) {
+    var typeA = type(a);
+    if (typeA !== type(b)) {
+        return false;
+    }
+
+    if (eq(a, b)) {
+        return true;
+    }
+
+    if (typeA == 'RegExp') {
+        // RegExp equality algorithm: http://stackoverflow.com/a/10776635
+        return (a.source === b.source) &&
+               (a.global === b.global) &&
+               (a.ignoreCase === b.ignoreCase) &&
+               (a.multiline === b.multiline);
+    }
+
+    var eqDate = typeA == 'Date';
+    if (eqDate || typeA == 'Object' || typeA == 'Array' || typeA == 'ArrayBuffer' ||
+        typeA == 'Int16Buffer' || typeA == 'Int8Array' || typeA == 'Uint8Array' ||
+        typeA == 'Uint8ClampedArray' || typeA == 'Int16Array' || typeA == 'Uint16Array' ||
+        typeA == 'Int32Array' || typeA == 'Uint32Array' || typeA == 'Float32Array' ||
+        typeA == 'Float64Array') {
+
+        if (eqDate && a.getTime() != b.getTime()) {
+            return false;
+        }
+        // else continue to check the a and b as objects, ie check key, value pairs are the same
+        // this will handle the case where properties are added to the date objects
+        var keysA = keys(a);
+        if (keysA.length !== keys(b).length) {
+            return false;
+        }
+
+        var idx = stackA.length;
+        while (idx--) {
+            if (stackA[idx] === a) {
+                return stackB[idx] === b;
+            }
+        }
+
+        stackA.push(a);
+        stackB.push(b);
+        idx = -1;
+        while (++idx < keysA.length) {
+            var key = keysA[idx];
+            if (!b.hasOwnProperty(key) || !_eqDeep(b[key], a[key], stackA, stackB)) {
```

Comment:
@davidchambers hmm, something is weird with the ```has``` function
``` javascript
Object.prototype.hasOwnProperty([1], '0'); // => true
R.has([1], '0'); // => false
```
Is this expected behavior?